File: 2023/10/10.py
from sys import stdin
from collections import deque

lines = stdin.read().strip().split('\n')

dirs = [(0, 1), (1, 0), (-1, 0), (0, -1)]
allowed = ["-7J", "|JL", "|F7", "-FL"]

def is_valid(y, x):
    return -0.5 <= y < len(lines) and -0.5 <= x < len(lines[0])

q = deque()
for y, line in enumerate(lines):
    if 'S' in line:
        q.append((y, line.index('S'), 0))

# Part 1: use BFS with allow_from and allow_to to check whether there is a connection
inside = {(y, x) for y in range(len(lines)) for x in range(len(lines[0]))}
visited = set()
max_dist = 0

while q:
    y, x, dist = q.popleft()
    max_dist = max(max_dist, dist)
    visited.add((y, x))
    inside.discard((y, x))

    for (dy, dx), allow_to, allow_from in zip(dirs, allowed, allowed[::-1]):
        ya, xa = y+dy, x+dx
        if is_valid(ya, xa) and lines[ya][xa] in allow_to and lines[y][x] in allow_from+"S":
            if (ya, xa) not in visited:
                q.append((ya, xa, dist+1))
            visited.add((y+dy/2, x+dx/2))


# Part 2: use a 0.5 grid to find all non-visited spaces, using the same visited set as in P1.
# Since we start outside of the grid, any cells inside will not be reached.
q.append((-0.5, -0.5))

while q:
    y, x = q.popleft()
    inside.discard((y, x))

    for dy, dx in dirs:
        ya, xa = y+dy/2, x+dx/2
        if is_valid(ya, xa) and (ya, xa) not in visited:
            q.append((ya, xa))
            visited.add((ya, xa))

print(max_dist)
print(len(inside))

File: 2023/10/10.rs
use std::{io::{stdin, Read}, collections::{VecDeque, HashSet}};

use itertools::{Itertools, izip, iproduct};

const DIR: [(i32, i32); 4] = [(0, 1), (1, 0), (-1, 0), (0, -1)];
const ALLOW_TO: [&str; 4] = ["-7J", "|JL", "|F7", "-FL"];
const ALLOW_FROM: [&str; 4] = ["-FLS", "|F7S", "|JLS", "-7JS"];

fn in_bounds(y: i32, x: i32, ys: usize, xs: usize) -> Option<(usize, usize)> {
    match (usize::try_from(y), usize::try_from(x)) {
        (Ok(y2), Ok(x2)) if y2 < ys && x2 < xs => Some((y2, x2)),
        _ => None,
    }
}

fn main() {
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let lines: Vec<&str> = input.lines().collect_vec();
    let (ys, xs) = (lines.len(), lines.get(0).unwrap().len());

    let start = input.lines()
        .find_position(|line| line.contains('S'))
        .map(|o| (o.0, o.1.find('S').unwrap(), 0));

    let mut q = VecDeque::from([start.unwrap()]);
    println!("{:?}", q);

    let mut visited = HashSet::new();
    let at = |y: usize, x: usize| lines[y].chars().nth(x).unwrap();
    let mut inside: HashSet<(usize, usize)> = iproduct!(0..ys, 0..xs).collect();

    let mut max_dist = 0;
    while !q.is_empty() {
        let (y, x, dist) = q.pop_front().unwrap();
        visited.insert((y+y+1, x+x+1));
        inside.remove(&(y, x));
        max_dist = max_dist.max(dist);
        for ((dy, dx), to, from) in izip!(DIR, ALLOW_TO, ALLOW_FROM) {
            if let Some((ya, xa)) = in_bounds(y as i32+dy, x as i32+dx, ys, xs) {
                if to.contains(at(ya, xa)) && from.contains(at(y, x)) {
                    if !visited.contains(&(y+ya+1, x+xa+1)) {
                        visited.insert((y+ya+1, x+xa+1));
                        q.push_back((ya, xa, dist+1));
                    }
                }
            }
        }
    }
    q.push_back((0, 0, 0));

    while !q.is_empty() {
        let (y, x, _) = q.pop_front().unwrap();
        visited.insert((y, x));
        if let Some((ydiv2, xdiv2)) = in_bounds((y as i32-1)/2, (x as i32-1)/2, ys, xs) {
            inside.remove(&(ydiv2, xdiv2));
        }
        for (dy, dx) in DIR {
            if let Some((ya, xa)) = in_bounds(y as i32+dy, x as i32+dx, ys*2+1, xs*2+1) {
                if !visited.contains(&(ya, xa)) {
                    visited.insert((ya, xa));
                    q.push_back((ya, xa, 0));
                }
            }
        }
    }

    println!("{}\n{}", max_dist, inside.len());
}